### DISCUSSION TITLE
Check Operators should raise AirflowFailException on check failure

### DISCUSSION THREAD

**Discussion by notatallshaw-gts** [apache/airflow#26712] [Ideas] [answered]
I need to add a retry on a task that uses `SQLValueCheckOperator` in case something unexpected happens (process fails, db connection doesn't work, or whatever).

However I don't want the task to retry if the result fails to pass the check, however as I understand the code right now it will because it raises `AirflowException`.

Would it make sense to change this to `AirflowFailException`  instead so it doesn't attempt to retry? And moreover would it be worth going through any similar "check Operators" for this behavior and make this the default guidance?

**Comment by potiuk** [comment]
I tihnk there should be a flag to decide what to do.  But yeah - feel free to make PRs for those.

**Reply by josh-fell** [reply]
cc @denimalpaca

**Comment by denimalpaca** [comment, marked as answer]
I'm working on a PR right now that is going to do some bug fixes across the board, so I can add this in as well.

**Reply by notatallshaw-gts** [reply]
Hey I see you edited your post to remove the `retries=0` suggestion, I assume you understand now why I want the task to retry in case of some external error but not retry when the check fails?

**Reply by denimalpaca** [reply]
Yep. I skimmed the q the first time, then re-read. I think it's a good flag to have. Something like `retry_on_check_fail`

**Reply by notatallshaw-gts** [reply]
This is implemented now via #26761 thanks!

**Comment by denimalpaca** [comment]
Would the correct behavior here be that if a record returns no values, that there also would be no retry? I would think so, just want to confirm.

**Reply by notatallshaw-gts** [reply]
My interpretation of having a check operator which has a flag  `retry_on_check_fail` that is set to `False` is if there is no value returned from the check then that is a "fail" of the check and should therefore not retry.

If someone is waiting for there to be a value they should be using a sensor (specifically in this case `SqlSensor`).

**Reply by denimalpaca** [reply]
Ok, so the two conditions where the `retry_on_check_fail` parameter would not allow retries are:
- No records are returned from the database
- One or more tests fail

This will be reflected in the PR associated with the issue.
